datumaro/plugins/open_images_format.py

```diff
@@ -0,0 +1,174 @@
+        except FileNotFoundError:
+            return
+
+        def set_parents_from_node(node, category):
+            for child_node in node.get('Subcategory', []):
+                _, child_category = label_categories.find(child_node['LabelName'])
+
+                if category is not None and child_category is not None:
+                    child_category.parent = category.name
+
+                set_parents_from_node(child_node, child_category)
+
+        _, root_category = label_categories.find(root_node['LabelName'])
+        set_parents_from_node(root_node, root_category)
+
+    def _load_items(self):
+        image_paths_by_id = {
+            osp.splitext(osp.basename(path))[0]: path
+            for path in find_images(
+                osp.join(self._dataset_dir, 'images'),
+                recursive=True, max_depth=1)
+        }
+
+        items_by_id = {}
+
+        def load_from(annotation_name):
+            with self._open_csv_annotation(annotation_name) as image_reader:
+                for image_description in image_reader:
+                    image_id = image_description['ImageID']
+                    if image_id in items_by_id:
+                        raise RepeatedItemError(item_id=image_id)
+
+                    subset = image_description['Subset']
+                    assert subset in {'test', 'train', 'validation'}
+
+                    items_by_id[image_id] = DatasetItem(
+                        id=image_id,
+                        image=image_paths_by_id.get(image_id),
+                        subset=subset,
+                    )
+
+        # It's preferable to load the combined image description file,
+        # because it contains descriptions for training images without human-annotated labels
+        # (the file specific to the training set doesn't).
+        # However, if it's missing, we'll try loading subset-specific files instead, so that
+        # this extractor can be used on individual subsets of the dataset.
+        try:
+            load_from(OpenImagesPath.FULL_IMAGE_DESCRIPTION_NAME)
+        except FileNotFoundError:
+            for pattern in OpenImagesPath.SUBSET_IMAGE_DESCRIPTION_PATTERNS:
+                for path in self._glob_annotations(pattern):
+                    load_from(path)
+
+        self._items.extend(items_by_id.values())
+
+        self._load_labels(items_by_id)
+
+    def _load_labels(self, items_by_id):
+        label_categories = self._categories[AnnotationType.label]
+
+        # TODO: implement reading of machine-annotated labels
+
+        for label_path in self._glob_annotations('*-human-imagelabels.csv'):
+            with self._open_csv_annotation(label_path) as label_reader:
+                for label_description in label_reader:
+                    image_id = label_description['ImageID']
+                    item = items_by_id[image_id]
+
+                    confidence = float(label_description['Confidence'])
+
+                    if 0.5 < confidence:
```

Comment:
Is it a dataset policy to ignore items with low confidence?

Reply:
A confidence below 1/2 indicates that the item is more likely to _not_ belong to the corresponding class than it is to belong to it. In fact, the human-verified label files contain "negative labels", which have a confidence of 0 (when an image definitely doesn't belong to a certain class).

The Datumaro `Label` class, however, doesn't have a confidence measure. If you just load all labels and discard the confidence information, it will not be possible to distinguish positive labels from negative labels, which will make the label annotations effectively useless.

I tried to extract as much information from the annotation files as is possible with the current Datumaro architecture, which meant only loading the positive labels. I think potentially Datumaro could be extended to support label confidence levels or negative labels, at which point it'll be possible to load all labels from OID.

Reply:
Arbitrary item and annotation attributes are currently stored in the `attributes` field of these types. Probably, you should add `confidence` as a `score` label attribute, to be consistent with other formats.

Reply:
OK, I did it. I'm assuming this means that I should no longer drop low-confidence (including 0 confidence) labels.